### ISSUE TITLE
Add warning if Node added to Editor UI has script not in Tool Mode

### ISSUE THREAD

**Issue by pycbouh** [godotengine/godot-proposals#520]
**Describe the project you are working on:**
Various editor plugins

**Describe the problem or limitation you are having in your project:**
It may be a matter of habit, but I find it easy to forget to add a `tool` instruction at the start of script files for Editor UI plugins. This then turns into an annoying problem, when there is no warning or error, and no output from your scripts either. The issue becomes somewhat obvious only in cases when you try to access specific methods defined in your script from another script.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
A warning in *Output* would be convenient to instantly indicate, that there is a script attached to your node, but it is not in Tool Mode, so it won't load while in Editor scene tree.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I'm not completely sure, what internal changes are required. There are two concerns though.

First, it should be noted, that not every script in `res://addons` should raise such warning. Some plugins provide nodes and scripts to be used in game, they are never loaded into Editor scene tree. Which is why I propose that validation is performed at runtime on changes to Editor scene tree, and not statically on contents of every script file in plugin folder.

Secondly, Tool Mode can also be activated in 2d/3d preview for any script. If previewed nodes are not sandboxed and are too added to Editor scene tree via some viewport, then some distinction should be determined. While there are several specific methods that plugins use to add UI (`add_control_to_dock`, `add_control_to_bottom_panel`, `add_control_to_container`, `add_tool_menu_item` and `add_tool_submenu_item`), sometimes nodes are added directly using calls like 
`get_editor_interface().get_editor_viewport().add_child`. This may make it harder to distinguish previewed nodes from plugin UI nodes, but should be handled correctly nonetheless.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I think it is possible to add a static check in some form via a plugin, but that's not ideal.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's a small enhancement for plugin creators, and should not require another plugin to be taken advantage of. Plus, ideal implementation requires changes to internal processes, I think.

**Comment by willnationsdev** [comment]
One thing that might help with this is that the editor can tell you which resources are dependent on and/or are owners of other resources. With this information, it would be easier to declare that all resources owned by an EditorPlugin script, and by extension, all resources which those resources own in turn, etc. must all be tool scripts if they are of type Script.

So, if you have an EditorPlugin that preloads a scene file into a const, and that scene file contains an external resource reference to a GDScript file cause one of its nodes has it attached, then the editor could produce a warning stating that the script isn't a tool script.

However, as I thought about this, I also realized that this would all cause problems if you wrote editor tools to, say, generate nodes in the node tree, but where those nodes themselves aren't supposed to have tool scripts.

In which case, the only way to accurately determine whether a given node is supposed to have a tool script or not is to check it directly within a method (such as the `add_control_to_*` methods mentioned in the OP) or to use some sort of function call to identify resources which should be tools. At *that* point though, it's basically no different from adding the `tool` keyword to the individual scripts anyway, so it still becomes up to the user to determine everything.

So, as much as I like the idea, I don't really see any other methods of doing this *except* for those methods specified, and even then, they'd only really be able to check the script on the nodes they are actually adding.

**Comment by pycbouh** [comment]
@willnationsdev If we take a look at the core of the problem, it boils down to the fact, that we have two distinct cases for tool scripts with no ability to distinguish between them. There are tool scripts used for editor extension by plugins and then there are tool scripts used for 2d/3d previews. When writing plugins, I find it undesirable to have every line of each script to run when I open a scene in the editor. It causes all sorts of problems and I have had to write around these problems a couple of times already, making my plugins less user friendly. So there may be a merit in creating any form of separation between these two cases and thus allowing for easy checks if plugin scripts are in correct state.

In other words, it is very concerning, if not only users cannot separate the two cases, but the engine cannot either. Not only does it make harder to write some plugins, it also makes it impossible to have plugin-specific sanity checks. So, if nothing else, a new way to denote plugin scripts must be introduced in 4.x.

**Comment by nonunknown** [comment]
This should be considered for 4.x
